# Bioinformatics - Smith-Waterman Alignment
Task 1 - Implement Smith-Waterman
Task 2 - Utilise Task 1 implementation to create a merged sequence from a set of local alignments

# Task 1
Task 1 implements Smith-Waterman local alignment, reading in 2 lines at a time from a given file and locally aligning those sequences in turn.

To run the code in terminal, navigate to the directory of the submission and enter:
```
    python Task1.py <input file name>
```


# Task 2
Task 2 implements reads in a template sequence and n many subsequent sequences, locally aligns those n sequences against the template sequence, and the constructs a new sequence from those n sequences based on their local alignment order.

```
    python Task2.py <input file name>
```
